crates/ruff_linter/src/suppression.rs

```diff
@@ -0,0 +1,905 @@
+use core::fmt;
+use ruff_python_parser::{TokenKind, Tokens};
+use ruff_source_file::LineRanges;
+use std::{error::Error, fmt::Formatter};
+use thiserror::Error;
+
+use ruff_python_trivia::{Cursor, is_python_whitespace};
+use ruff_text_size::{Ranged, TextLen, TextRange, TextSize, TextSlice};
+use smallvec::{SmallVec, smallvec};
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+enum SuppressionAction {
+    Disable,
+    Enable,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub(crate) struct SuppressionComment {
+    /// Range containing the entire suppression comment
+    range: TextRange,
+
+    /// How indented an own-line comment is, or None for trailing comments
+    indent: Option<String>,
+
+    /// The action directive
+    action: SuppressionAction,
+
+    /// Ranges containing the lint codes being suppressed
+    codes: SmallVec<[TextRange; 2]>,
+
+    /// Range containing the reason for the suppression
+    reason: TextRange,
+}
+
+impl SuppressionComment {
+    /// Return the suppressed codes as strings
+    fn codes_as_str(&self, source: &str) -> Vec<String> {
+        self.codes
+            .iter()
+            .map(|range| source.slice(range).to_string())
+            .collect()
+    }
```

Comment:
Allocating a `Vec` just for comparing the code and copying over the string slices is unnecessarily expensive. 

You can do something like this


```suggestion
    fn codes_as_str<'src>(&self, source: &'src str) -> impl Iterator<Item = &'src str> {
        self.codes.iter().map(|range| source.slice(range))
    }
```

And then compare the codes using `comment.codes_as_str().eq(other.codes_as_str())`


However, this will require that the codes are in the exact same order. Is this the semantics we want?